api/v1alpha3/domachine_webhook.go

```diff
@@ -0,0 +1,96 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package v1alpha3
+
+import (
+	"reflect"
+
+	"github.com/pkg/errors"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	runtime "k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/util/validation/field"
+	ctrl "sigs.k8s.io/controller-runtime"
+	logf "sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/webhook"
+)
+
+// log is for logging in this package.
+var _ = logf.Log.WithName("domachine-resource")
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-infrastructure-cluster-x-k8s-io-v1alpha3-domachine,mutating=false,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=domachines,versions=v1alpha3,name=validation.domachine.infrastructure.cluster.x-k8s.io,sideEffects=None
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-infrastructure-cluster-x-k8s-io-v1alpha3-domachine,mutating=true,failurePolicy=fail,matchPolicy=Equivalent,groups=infrastructure.cluster.x-k8s.io,resources=domachines,versions=v1alpha3,name=default.domachine.infrastructure.cluster.x-k8s.io,sideEffects=None
+
+var (
+	_ webhook.Defaulter = &DOMachine{}
+	_ webhook.Validator = &DOMachine{}
+)
+
+func (r *DOMachine) SetupWebhookWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewWebhookManagedBy(mgr).
+		For(r).
+		Complete()
+}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (r *DOMachine) Default() {}
+
+// ValidateCreate implements webhook.Validator so a webhook will be registered for the type.
+func (r *DOMachine) ValidateCreate() error {
+	return nil
+}
+
+// ValidateUpdate implements webhook.Validator so a webhook will be registered for the type
+func (r *DOMachine) ValidateUpdate(old runtime.Object) error {
+	var allErrs field.ErrorList
+
+	newDOMachine, err := runtime.DefaultUnstructuredConverter.ToUnstructured(r)
+	if err != nil {
+		return apierrors.NewInternalError(errors.Wrap(err, "failed to convert new DOMachine to unstructured object"))
+	}
+
+	oldDOMachine, err := runtime.DefaultUnstructuredConverter.ToUnstructured(old)
+	if err != nil {
+		return apierrors.NewInternalError(errors.Wrap(err, "failed to convert new DOMachine to unstructured object"))
+	}
+
+	newDOMachineSpec := newDOMachine["spec"].(map[string]interface{})
+	oldDOMachineSpec := oldDOMachine["spec"].(map[string]interface{})
+
+	// allow changes to providerID
```

Comment:
AFAICS, the provider ID is set from the (read-only) droplet ID. What's the use case for it to be allowed to change? 🤔

Reply:
afaik, it required for the controller to set the providerID https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1218#discussion_r343751302 although, I also think if we can reject updating providerID only if it already set.

but can we leave it as is? and keeping an eye on the other providers on how they handle this. since it's not clearly documented yet what validations should be enforcing https://github.com/kubernetes-sigs/cluster-api-provider-aws/pull/1218#issuecomment-553446074

Reply:
Ah you're right, I missed the fact that the provider ID is set by the CAPDO controller itself which constitutes an update.

Disallowing updates once the provider ID is set sounds reasonable. Happy to keep things the way you propose them in your PR for now though and see what consensus is made with regards to webhook validations.